src/wasm/wasm-validator.cpp

```diff
@@ -559,6 +559,26 @@ void WasmValidator::visitFunction(Function *curr) {
   shouldBeTrue(breakTargets.empty(), curr->body, "all named break targets must exist");
   returnType = unreachable;
   labelNames.clear();
+  // expressions must not be seen more than once
+  struct Walker : public PostWalker<Walker, UnifiedExpressionVisitor<Walker>> {
+    std::unordered_set<Expression*>& seen;
+    std::vector<Expression*> dupes;
+
+    Walker(std::unordered_set<Expression*>& seen) : seen(seen) {}
+
+    void visitExpression(Expression* curr) {
+      if (seen.find(curr) == seen.end()) {
```

Comment:
Can you just do
`if (seen.insert(curr).second) dupes.push_back(curr);`

Reply:
I suppose, but I find the `.second` notation hard to read myself (a shame it's not `.inserted`). Is it just me?

Reply:
No, I agree the API is ugly.
Could always do something like
```
bool inserted;
std::tie<std::ignore, inserted> = seen.insert(curr);
if (!inserted) seen.insert(curr);
```

Reply:
Thanks, I like that way, updated.